File: leetcode/23/main.cpp
#include <queue>
#include <vector>
using namespace std;

struct ListNode {
  int val;
  ListNode *next;
  ListNode() : val(0), next(nullptr) {
  }
  ListNode(int x) : val(x), next(nullptr) {
  }
  ListNode(int x, ListNode *next) : val(x), next(next) {
  }
};

using piN = pair<int, ListNode *>;

class Solution {
public:
  ListNode *mergeKLists(vector<ListNode *> &lists) {
    priority_queue<piN, vector<piN>, greater<piN>> pq;
    for (auto *list : lists) {
      if (list) pq.push({ list->val, list });
    }

    ListNode head, *cur = &head;
    while (!pq.empty()) {
      auto [val, node] = pq.top();
      pq.pop();
      cur->next = node;
      cur = cur->next;
      if (node->next) {
        node = node->next;
        pq.push({ node->val, node });
      }
    }

    return head.next;
  }
};

File: leetcode/23/main.js
function ListNode(val, next) {
  this.val = val === undefined ? 0 : val;
  this.next = next === undefined ? null : next;
}

/**
 * @param {ListNode[]} lists
 * @return {ListNode}
 */
var mergeKLists = function (lists) {
  const pq = new MinPriorityQueue((node) => node.val);
  for (const node of lists) {
    if (node) pq.enqueue(node);
  }

  const head = new ListNode();
  let cur = head;

  while (!pq.isEmpty()) {
    const node = pq.dequeue();
    cur.next = node;
    cur = cur.next;
    if (node.next) pq.enqueue(node.next);
  }

  return head.next;
};
